### ISSUE TITLE
400 Bad Request error from DO Spaces  

### ISSUE THREAD

**Issue by derkgort** [s3ql/s3ql#250]
Hi,

First: Thanks for making an awesome tool!

I'm having an issue with DO spaces since a few days. They had some downtime in Amsterdam and since that happened s3ql is not keeping the mount mounted.
I have updated s3ql to the latest, but it it still happening.

What is happening is the following:
- I mount the s3c
- I start a backup that writes large files to the mount
- after a while, ~ 10-15 minutes it failes with a "400 Bad Request"
- the mount becomes unavailabe and i have to unmount, fsck and mount again with the backup failed.

I've set s3ql in Debug and this is the output before it breaks. It shows a "400 Bad Request" being suppressed:

mount.s3ql: 2021-06-09 09:26:04.304 31125 DEBUG    MainThread s3ql.block_cache.upload_if_dirty: started with <CacheEntry, inode=15553, blockno=226>
mount.s3ql: 2021-06-09 09:26:04.305 31125 DEBUG    MainThread s3ql.block_cache.expire: removing <CacheEntry, inode=15553, blockno=226> from cache
mount.s3ql: 2021-06-09 09:26:04.322 31125 DEBUG    MainThread s3ql.block_cache.upload_if_dirty: started with <Dirty CacheEntry, inode=15553, blockno=227>
mount.s3ql: 2021-06-09 09:26:04.322 31125 DEBUG    MainThread s3ql.block_cache.upload_if_dirty: started with <CacheEntry, inode=15554, blockno=94>
mount.s3ql: 2021-06-09 09:26:04.323 31125 DEBUG    MainThread s3ql.block_cache.expire: removing <CacheEntry, inode=15554, blockno=94> from cache
mount.s3ql: 2021-06-09 09:26:04.338 31125 DEBUG    MainThread s3ql.block_cache.upload_if_dirty: started with <Dirty CacheEntry, inode=15553, blockno=228>
mount.s3ql: 2021-06-09 09:26:04.339 31125 DEBUG    MainThread s3ql.block_cache.expire: waiting for transfer threads..
mount.s3ql: 2021-06-09 09:26:07.308 31125 DEBUG    Thread-2 s3ql.block_cache._do_upload: uploaded 104857600 bytes in 15.430 seconds, 6.48 MiB/s
mount.s3ql: 2021-06-09 09:26:07.309 31125 DEBUG    MainThread s3ql.block_cache.upload_if_dirty: started with <CacheEntry, inode=15553, blockno=227>
mount.s3ql: 2021-06-09 09:26:07.309 31125 DEBUG    MainThread s3ql.block_cache.expire: removing <CacheEntry, inode=15553, blockno=227> from cache
mount.s3ql: 2021-06-09 09:26:07.313 31125 DEBUG    Thread-2 s3ql.block_cache._upload_loop: reading from upload queue...
mount.s3ql: 2021-06-09 09:26:07.326 31125 DEBUG    MainThread s3ql.block_cache.upload_if_dirty: started with <Dirty CacheEntry, inode=15553, blockno=228>
mount.s3ql: 2021-06-09 09:26:07.326 31125 DEBUG    MainThread s3ql.block_cache.expire: waiting for transfer threads..
mount.s3ql: 2021-06-09 09:26:13.063 31125 DEBUG    Thread-3 s3ql.block_cache._do_upload: uploaded 87322624 bytes in 13.070 seconds, 6.37 MiB/s
mount.s3ql: 2021-06-09 09:26:13.065 31125 DEBUG    MainThread s3ql.block_cache.upload_if_dirty: started with <CacheEntry, inode=15553, blockno=228>
mount.s3ql: 2021-06-09 09:26:13.065 31125 DEBUG    MainThread s3ql.block_cache.expire: removing <CacheEntry, inode=15553, blockno=228> from cache
mount.s3ql: 2021-06-09 09:26:13.067 31125 DEBUG    Thread-3 s3ql.block_cache._upload_loop: reading from upload queue...
mount.s3ql: 2021-06-09 09:26:13.082 31125 DEBUG    MainThread s3ql.block_cache.expire: finished
mount.s3ql: 2021-06-09 09:26:13.082 31125 DEBUG    MainThread s3ql.block_cache.drop: finished
mount.s3ql: 2021-06-09 09:26:13.082 31125 DEBUG    MainThread s3ql.block_cache.destroy: Signaling upload threads...
mount.s3ql: 2021-06-09 09:26:13.083 31125 DEBUG    Thread-2 s3ql.block_cache._upload_loop: got QuitSentinel
mount.s3ql: 2021-06-09 09:26:13.084 31125 DEBUG    Thread-3 s3ql.block_cache._upload_loop: got QuitSentinel
mount.s3ql: 2021-06-09 09:26:13.085 31125 DEBUG    Thread-4 s3ql.block_cache._upload_loop: got QuitSentinel
mount.s3ql: 2021-06-09 09:26:18.090 31125 DEBUG    MainThread s3ql.block_cache.destroy: Signaling removal threads...
mount.s3ql: 2021-06-09 09:26:18.091 31125 DEBUG    MainThread s3ql.block_cache.destroy: waiting for upload threads...
mount.s3ql: 2021-06-09 09:26:18.101 31125 DEBUG    MainThread s3ql.block_cache.destroy: waiting for removal threads...
mount.s3ql: 2021-06-09 09:26:18.105 31125 DEBUG    MainThread s3ql.block_cache.destroy: cleanup done.
mount.s3ql: 2021-06-09 09:26:18.121 31125 DEBUG    MainThread s3ql.mount.exchook: Suppressing exception hook for re-raised 400 Bad Request
systemd: s3ql.service: main process exited, code=exited, status=1/FAILURE
systemd: Unit s3ql.service entered failed state.
systemd: s3ql.service failed.

Is there anything that stands out as the cause? I've tried to find a cause, but nothing.
I do have the "feeling" that is has something to do with deleting or syncing the DB to DigitalOcean spaces.

My config works without issues with AWS S3.

s3cmd works with DO too.

My guess is that they did some update to their software as there were migrating for days before it broke.
I've opened a ticket at DO, but well... Tickets. You know how that works..

Is there anything for me to try to get more information / details on where this is failing?

Thanks!

**Comment by Nikratio** [comment]
It's failing on the remote side, where the server claims that S3QL sends an invalid request. Since it works just fine with S3, it's likely that the problem is not with S3QL but with your storage provider.


The above log misses the most important part. Could you provide more context? You do not need to enable debug logging for this.

**Comment by derkgort** [comment]
This is the error from when it was running 3.6.0 without debug:

Jun  7 15:25:19 Server-name umount.s3ql: ERROR: Uncaught top-level exception:
Jun  7 15:25:19 Server-name umount.s3ql: Traceback (most recent call last):
Jun  7 15:25:19 Server-name umount.s3ql: File "/opt/s3ql-virtualenv/bin/umount.s3ql", line 11, in <module>
Jun  7 15:25:19 Server-name umount.s3ql: load_entry_point('s3ql==3.6.0', 'console_scripts', 'umount.s3ql')()
Jun  7 15:25:19 Server-name umount.s3ql: File "/opt/s3ql-virtualenv-3.6.0/lib64/python3.6/site-packages/s3ql-3.6.0-py3.6-linux-x86_64.egg/s3ql/umount.py", line 185, in main
Jun  7 15:25:19 Server-name umount.s3ql: assert_s3ql_mountpoint(options.mountpoint)
Jun  7 15:25:19 Server-name umount.s3ql: File "/opt/s3ql-virtualenv-3.6.0/lib64/python3.6/site-packages/s3ql-3.6.0-py3.6-linux-x86_64.egg/s3ql/common.py", line 232, in assert_s3ql_mountpoint
Jun  7 15:25:19 Server-name umount.s3ql: ctrlfile = assert_s3ql_fs(mountpoint)
Jun  7 15:25:19 Server-name umount.s3ql: File "/opt/s3ql-virtualenv-3.6.0/lib64/python3.6/site-packages/s3ql-3.6.0-py3.6-linux-x86_64.egg/s3ql/common.py", line 197, in assert_s3ql_fs
Jun  7 15:25:19 Server-name umount.s3ql: if not (CTRL_NAME not in pyfuse3.listdir(path)
Jun  7 15:25:19 Server-name umount.s3ql: File "src/pyfuse3.pyx", line 526, in pyfuse3.listdir
Jun  7 15:25:19 Server-name umount.s3ql: OSError: [Errno 107] Transport endpoint is not connected: '/XXXXXXXXXXXx'
Jun  7 15:25:19 Server-name systemd: s3ql.service: control process exited, code=exited status=1
Jun  7 15:25:19 Server-name systemd: Unit s3ql.service entered failed state.
Jun  7 15:25:19 Server-name systemd: s3ql.service failed.


This is from 3.7.3 without debug:

Jun  8 11:27:25 Server-name systemd: Started mount s3ql filesystem.
Jun  8 11:41:17 Server-name mount.s3ql: Unhandled exception in thread, terminating
Jun  8 11:41:17 Server-name mount.s3ql: Traceback (most recent call last):
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/mount.py", line 58, in run_with_except_hook
Jun  8 11:41:17 Server-name mount.s3ql: run_old(*args, **kw)
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/rh/rh-python38/root/usr/lib64/python3.8/threading.py", line 870, in run
Jun  8 11:41:17 Server-name mount.s3ql: self._target(*self._args, **self._kwargs)
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/block_cache.py", line 319, in _upload_loop
Jun  8 11:41:17 Server-name mount.s3ql: self._do_upload(*tmp)
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/block_cache.py", line 384, in _do_upload
Jun  8 11:41:17 Server-name mount.s3ql: obj_size = backend.perform_write(do_write, 's3ql_data_%d'
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/backends/common.py", line 108, in wrapped
Jun  8 11:41:17 Server-name mount.s3ql: return method(*a, **kw)
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/backends/common.py", line 279, in perform_write
Jun  8 11:41:17 Server-name mount.s3ql: return fn(fh)
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/backends/comprenc.py", line 554, in __exit__
Jun  8 11:41:17 Server-name mount.s3ql: self.close()
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/backends/comprenc.py", line 548, in close
Jun  8 11:41:17 Server-name mount.s3ql: self.fh.close()
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/backends/common.py", line 108, in wrapped
Jun  8 11:41:17 Server-name mount.s3ql: return method(*a, **kw)
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/backends/s3c.py", line 912, in close
Jun  8 11:41:17 Server-name mount.s3ql: resp = self.backend._do_request('PUT', '/%s%s' % (self.backend.prefix, self.key),
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/backends/s3c.py", line 531, in _do_request
Jun  8 11:41:17 Server-name mount.s3ql: self._parse_error_response(resp)
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/backends/s3c.py", line 553, in _parse_error_response
Jun  8 11:41:17 Server-name mount.s3ql: raise HTTPError(resp.status, resp.reason, resp.headers)
Jun  8 11:41:17 Server-name mount.s3ql: s3ql.backends.s3c.HTTPError: 400 Bad Request
Jun  8 11:41:17 Server-name mount.s3ql: Uncaught top-level exception:
Jun  8 11:41:17 Server-name mount.s3ql: Traceback (most recent call last):
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/mount.py", line 58, in run_with_except_hook
Jun  8 11:41:17 Server-name mount.s3ql: run_old(*args, **kw)
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/rh/rh-python38/root/usr/lib64/python3.8/threading.py", line 870, in run
Jun  8 11:41:17 Server-name mount.s3ql: self._target(*self._args, **self._kwargs)
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/block_cache.py", line 319, in _upload_loop
Jun  8 11:41:17 Server-name mount.s3ql: self._do_upload(*tmp)
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/block_cache.py", line 384, in _do_upload
Jun  8 11:41:17 Server-name mount.s3ql: obj_size = backend.perform_write(do_write, 's3ql_data_%d'
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/backends/common.py", line 108, in wrapped
Jun  8 11:41:17 Server-name mount.s3ql: return method(*a, **kw)
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/backends/common.py", line 279, in perform_write
Jun  8 11:41:17 Server-name mount.s3ql: return fn(fh)
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/backends/comprenc.py", line 554, in __exit__
Jun  8 11:41:17 Server-name mount.s3ql: self.close()
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/backends/comprenc.py", line 548, in close
Jun  8 11:41:17 Server-name mount.s3ql: self.fh.close()
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/backends/common.py", line 108, in wrapped
Jun  8 11:41:17 Server-name mount.s3ql: return method(*a, **kw)
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/backends/s3c.py", line 912, in close
Jun  8 11:41:17 Server-name mount.s3ql: resp = self.backend._do_request('PUT', '/%s%s' % (self.backend.prefix, self.key),
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/backends/s3c.py", line 531, in _do_request
Jun  8 11:41:17 Server-name mount.s3ql: self._parse_error_response(resp)
Jun  8 11:41:17 Server-name mount.s3ql: File "/opt/s3ql-virtualenv-3.7.3/lib64/python3.8/site-packages/s3ql-3.7.3-py3.8-linux-x86_64.egg/s3ql/backends/s3c.py", line 553, in _parse_error_response
Jun  8 11:41:17 Server-name mount.s3ql: raise HTTPError(resp.status, resp.reason, resp.headers)
Jun  8 11:41:17 Server-name mount.s3ql: s3ql.backends.s3c.HTTPError: 400 Bad Request
Jun  8 11:41:17 Server-name mount.s3ql: Unmounting file system...
Jun  8 11:42:28 Server-name systemd: s3ql.service: main process exited, code=exited, status=1/FAILURE
Jun  8 11:42:28 Server-name systemd: Unit s3ql.service entered failed state.
Jun  8 11:42:28 Server-name systemd: s3ql.service failed.

**Comment by Nikratio** [comment]
Yeah, no surprises here. Your storage server ("Do Spaces" I asssume?) claims that S3QL is sending invalid PUT requests.

**Comment by derkgort** [comment]
Ok,
I'll point Digital Ocean to that. But i've given up hope that they are going to fix it. Because s3cmd is working without any problems, and S3QL / "block storage" is not an officially supported use case.

Thanks for helping with this.

**Comment by robbat2** [comment]
:wave: Hi, I'm one of the DigitalOcean engineers on the Spaces product.
We absolutely did upgrade some software as part of working on the outage in the Amsterdam region.

Question to the S3QL developers:
Could you show users how to capture the complete HTTP request+response headers when they get a `400 Bad Request` response, so they can provide them on the support ticket?

If it uses keep-alive connections, then the request+response headers on the previous request in the same connection as well would be valuable.

The logging on the side of Spaces doesn't retain every single one of the headers, as the vast majority of the `400 Bad Request` responses are due to customers writing their own S3 clients and sending some malformed requests.

**Comment by Nikratio** [comment]
S3QL uses dugong, so I think the easiest way is to use the `DUGONG_TRACEFILE` environment variable. Before starting mount.s3ql (or fsck.s3ql), run `export DUGONG_TRACEFILE=/tmp/http_capture_%d.txt`. This will produce a series of files (with `%d` replaced by a number) that contain the *complete* request and response, INCLUDING PASSWORDS AND EVERYTHING.

Source: https://github.com/python-dugong/python-dugong/blob/master/dugong/__init__.py#L549

**Comment by robbat2** [comment]
Is that going to dump the request+response body payloads as well?

If the are S3 calls w/ XML request payloads, they might be relevant in other cases, but in this specific case, I can see in the logs we do have that the `400 Bad Request` was on `s3:DeleteObject`,`s3:GetObject`,` s3:PutObject` so we don't need them, esp if they are going to make the tracefiles much larger.

**Comment by robbat2** [comment]
@derkgort still waiting for you to post those debug logs to your DigitalOcean support ticket, so we can trace it down better.

**Comment by Nikratio** [comment]
I'll close this bug for now, since to me it looks as if DigitalOcean isn't 100% S3 compatible and refusing what AWS S3 considers valid requests.

**Comment by derkgort** [comment]
I can concur. I have migrated all to B2 in the meantime.
They are still working on it, but there is not a lot of feedback.
It seems they broke something that made some objects unavailable, while still visible.
I'm happy to report that I was able to copy everything I had stored by making 2 s3ql mounts and rsyncing it from DO to B2. ( It did take a week... mostly because of the many timeouts from DO )

S3QL is great! Thanks for the effort of making this tool.
